### PR TITLE
JVM_IR: recover accidentally lost backingField (KT-49998)

### PR DESCRIPTION
**Pull request by scaventz**
issue: [KT-49998](https://youtrack.jetbrains.com/issue/KT-49998)

### CODE REVIEW

**Comment by Mtan6666666** [comment]
Trust wallet

File: compiler/testData/codegen/box/annotations/annotationDefaultValueOfUnsigned.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+annotation class Ann4(val value: ULong = 44u)
+
+@Ann1
+@Ann2
+@Ann3
+@Ann4
+class A
+
+fun box(): String {
+    val default1 = A::class.java.getDeclaredAnnotation(Ann1::class.java).value
```
Comment (udalov) [inline compiler/testData/codegen/box/annotations/annotationDefaultValueOfUnsigned.kt:18]: `Class.getDeclaredAnnotation` is missing on JDK6 and older versions of Android, so the corresponding tests (`compiler/tests-different-jdk`, `compiler/android-tests`) are failing. Let's use `Class.getAnnotation` instead

**Comment by udalov** [comment]
The change feels strange because we already don't mangle annotation property accessors. It's just `value` in the bytecode, without the mangling suffix. Which hints that the root problem might be located elsewhere.

I remember similar problems with other cases when corresponding property of accessors (`IrSimpleFunction.correspondingProperyt`) was accidentally lost during IR transformation. Does it seem like something that's happening here? In that case, it'd be better to track where it's being lost and fix that instead, because it may cause other problems.

File: compiler/testData/codegen/box/annotations/annotationDefaultValueOfUnsigned.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+annotation class Ann1(val value: UByte = 41u)
+annotation class Ann2(val value: UShort = 42u)
+annotation class Ann3(val value: UInt = 43u)
+annotation class Ann4(val value: ULong = 44u)
+
+@Ann1
+@Ann2
+@Ann3
+@Ann4
+class A
```
Comment (ilmirus) [inline compiler/testData/codegen/box/annotations/annotationDefaultValueOfUnsigned.kt:15]: Can you, please, also test an inline class other than unsigned integral type?

**Review by ilmirus — CHANGES_REQUESTED** [review]

File: compiler/testData/codegen/box/annotations/annotationDefaultValueOfUnsigned.kt (hunk 2 above)
Comment (ilmirus) [reply compiler/testData/codegen/box/annotations/annotationDefaultValueOfUnsigned.kt:15]: Ah, nevermind. Usual inline classes are not allowed inside annotation.

**Review by ilmirus — APPROVED** [review]

**Comment by udalov** [comment]
The change looks good to me too, but I need some time to investigate what exactly is the issue with inline classes compared to non-inline classes.

**Comment by scaventz** [comment]
Sorry for not providing detailes ealier.
As far as I know, in `MemoizedInlineClassReplacements:78`, because the inline return type of getValue meets the condition

```kotlin
val IrFunction.hasMangledReturnType: Boolean
    get() = returnType.isInlineClassType() && parentClassOrNull?.isFileClass != true
```

, the `backingField` get accidentally lost in `createMethodReplacement` method (used to compute default value of annotation in `getAnnotationDefaultValueExpression` method of `FunctionCodegen`).

It seems the method `createMethodReplacement` is mainly just mangle the name of function `getValue` and maybe unnecessary, because in the end the mangled name is never adopted( see method `mapSignature`of `MethodSignatureMapper.kt`)
But even if the mangle process can be avoid here, I agree to do a simple and intuitive fix first, like this PR does.

**Comment by udalov** [comment]
Thanks for the explanation, it makes sense and indeed seems like the simplest fix. Merged to master, will be released in 1.6.20. Thank you for the contribution!

**Pull request merged**